src/threadpool.c

```diff
@@ -57,104 +59,132 @@ static void uv__cancelled(struct uv__work* w) {
 }
 
 
+/* Only call when you hold NO locks. */
+static void do_work(struct uv__work* w) {
+  w->work(w);
+  uv_mutex_lock(&w->loop->wq_mutex);
+  w->work = NULL;  /* Signal uv_cancel() that the work req is done executing. */
+  QUEUE_INSERT_TAIL(&w->loop->wq, &w->wq);
+  uv_async_send(&w->loop->wq_async);
+  uv_mutex_unlock(&w->loop->wq_mutex);
+}
+
+
 /* To avoid deadlock with uv_cancel() it's crucial that the worker
  * never holds the global mutex and the loop-local mutex at the same time.
  */
-static void worker(void* arg) {
-  struct uv__work* w;
+#ifdef _WIN32
+static unsigned __stdcall worker(void* arg) {
+#else
+static void* worker(void* arg) {
+#endif
   QUEUE* q;
 
-  (void) arg;
+  (void) &arg;
+  uv_mutex_lock(&mutex);
 
   for (;;) {
-    uv_mutex_lock(&mutex);
+    if (QUEUE_EMPTY(&wq))
+      uv_cond_timedwait(&cond, &mutex, IDLE_THREAD_TIMEOUT);
 
-    while (QUEUE_EMPTY(&wq))
-      uv_cond_wait(&cond, &mutex);
+    if (QUEUE_EMPTY(&wq))
+      break;  /* Timed out, nothing to do, exit thread. */
 
     q = QUEUE_HEAD(&wq);
 
-    if (q == &exit_message)
-      uv_cond_signal(&cond);
-    else {
-      QUEUE_REMOVE(q);
-      QUEUE_INIT(q);  /* Signal uv_cancel() that the work req is
-                             executing. */
+    if (q == &exit_message) {
+      uv_cond_signal(&cond);  /* Propagate exit message to next thread. */
+      break;
     }
 
-    uv_mutex_unlock(&mutex);
+    QUEUE_REMOVE(q);
+    QUEUE_INIT(q);  /* Signal uv_cancel() that the work req is executing. */
 
-    if (q == &exit_message)
-      break;
+    busy_threads += 1;
+    uv_mutex_unlock(&mutex);
+    do_work(QUEUE_DATA(q, struct uv__work, wq));
+    uv_mutex_lock(&mutex);
+    busy_threads -= 1;
+  }
 
-    w = QUEUE_DATA(q, struct uv__work, wq);
-    w->work(w);
+  assert(num_threads > INT_MIN);  /* Extremely paranoid sanity check. */
+  num_threads -= 1;
+  uv_mutex_unlock(&mutex);
 
-    uv_mutex_lock(&w->loop->wq_mutex);
-    w->work = NULL;  /* Signal uv_cancel() that the work req is done
-                        executing. */
-    QUEUE_INSERT_TAIL(&w->loop->wq, &w->wq);
-    uv_async_send(&w->loop->wq_async);
-    uv_mutex_unlock(&w->loop->wq_mutex);
-  }
+  return 0;
 }
 
 
-static void post(QUEUE* q) {
-  uv_mutex_lock(&mutex);
-  QUEUE_INSERT_TAIL(&wq, q);
-  uv_cond_signal(&cond);
-  uv_mutex_unlock(&mutex);
+static int new_detached_thread(void) {
+#ifdef _WIN32
+  HANDLE thread;
+
+  thread = (HANDLE) _beginthreadex(NULL, 0, worker, NULL, 0, NULL);
+  if (thread == NULL)
+    return -errno;
+
+  if (CloseHandle(thread))
+    abort();
+
+  return 0;
+#else
+  pthread_t thread;
+  int err;
+
+  err = pthread_create(&thread, NULL, worker, NULL);
+  if (err)
+    return -err;
+
+  if (pthread_detach(thread))
+    abort();
+
+  return 0;
+#endif
 }
```

Comment:
Can be removed if/when libuv grows a `uv_thread_detach()` function.  I can do that in a follow-up PR.